### PR TITLE
[WIP] Default resource List page to use additionalPrinterColumns of the resource

### PR DESCRIPTION
**Pull request by jhadvig**
/assign @spadgett 

Not really sure if the calculating the column size is proper, thats why Im putting the PR as WIP and will give it a bit more thoughts.

### CODE REVIEW

**Comment by spadgett** [comment]
@smarterclayton @jwforres FYI, this adds server printing for list pages

File: frontend/public/ui/ui-actions.js [hunk 1]
```diff
@@ -160,6 +164,17 @@ export const UIActions = {
 
   [types.setCreateProjectMessage]: message => ({type: types.setCreateProjectMessage, message}),
 
+  [types.setTableObject]: (crdModel) => dispatch => {
```
Comment (spadgett) [inline frontend/public/ui/ui-actions.js]: This really feels like it belongs in k8s-actions / k8s-reducers. Shouldn't we generate a unique ID like `watchK8sList` etc?

File: frontend/public/components/factory/list-page.jsx [hunk 2]
```diff
@@ -57,6 +59,14 @@ TextFilter.displayName = 'TextFilter';
 
 /** @augments {React.PureComponent<{ListComponent: React.ComponentType<any>, kinds: string[], flatten?: function, data?: any[], rowFilters?: any[]}>} */
 export class ListPageWrapper_ extends React.PureComponent {
+  componentDidMount() {
+    if (_.size(this.props.kinds) === 1) {
+      store.dispatch(UIActions.setTableObject(kindObj(this.props.kinds[0])));
```
Comment (spadgett) [inline frontend/public/components/factory/list-page.jsx]: We should only request the table output for pages that need it (default list pages with no custom table columns).

We also shouldn't be watching resources in the default table since we're using the table data. It looks like we're opening a watch and not using the data.

File: frontend/public/components/default-resource.jsx [hunk 3]
```diff
@@ -17,30 +18,65 @@ import {
 const { common } = Cog.factory;
 const menuActions = [...common];
 
-const Header = props => <ListHeader>
-  <ColHead {...props} className="col-xs-6 col-sm-4" sortField="metadata.name">Name</ColHead>
-  <ColHead {...props} className="col-xs-6 col-sm-4" sortField="metadata.namespace">Namespace</ColHead>
-  <ColHead {...props} className="col-sm-4 hidden-xs" sortField="metadata.creationTimestamp">Created</ColHead>
-</ListHeader>;
+const computeColumnSizes = (columns, isResourceNamespaced) => {
```
Comment (spadgett) [inline frontend/public/components/default-resource.jsx]: I'd say we want a maximum number of columns to show at each screen width breakpoint and drop columns with lower priority.

Alternately, we could make the table horizontally scrollable.

File: frontend/public/components/default-resource.jsx [hunk 4]
```diff
@@ -17,30 +18,65 @@ import {
...
+  if (!_.get(props, ['resourceTable', 'table'])) {
+    return null;
+  }
+  const { table, isNamespaced } = props.resourceTable;
+  const { fistColumnSize, columnsSize } = computeColumnSizes(table.columnDefinitions, isNamespaced);
+  const columnHeaders = _.map(table.columnDefinitions, (column, index) => {
+    if (index === 0) {
+      return <ColHead key={index} className={`col-xs-${fistColumnSize}`}>{column.name}</ColHead>;
+    }
+    return <ColHead key={isNamespaced ? index + 1 : index} className={`col-xs-${columnsSize}`}>{column.name}</ColHead>;
```
Comment (spadgett) [inline frontend/public/components/default-resource.jsx]: We'll eventually need to make sorting work.

File: frontend/public/components/default-resource.jsx [hunk 5]
```diff
@@ -17,30 +18,65 @@ import {
...
+  const namespace = _.get(row, 'object.metadata.namespace');
+  const { fistColumnSize, columnsSize } = computeColumnSizes(cells, isNamespaced);
+  const renderedRows = _.map(cells, (column, i) => {
+    if (i === 0 ) {
+      return <div key={i} className={`col-xs-${fistColumnSize} co-resource-link-wrapper`}>
+        <ResourceCog actions={menuActions} kind={referenceFor(obj) || kind} resource={obj} />
+        <ResourceLink kind={kind} name={obj.metadata.name} namespace={obj.metadata.namespace} title={obj.metadata.name} />
+      </div>;
+    }
+    return <div key={isNamespaced ? i + 1 : i} className={`col-xs-${columnsSize} co-break-word`}>{column}</div>;
```
Comment (spadgett) [inline frontend/public/components/default-resource.jsx]: We need to make sure we're formatting the column based on the type. Dates and possibly other types should be rendered specially instead of using the raw value that's returned.

File: frontend/public/components/factory/list-page.jsx (hunk 2 above)
Comment (spadgett) [reply frontend/public/components/factory/list-page.jsx]: We probably want to add polling to this eventually.

File: frontend/public/ui/ui-actions.js (hunk 1 above)
Comment (jhadvig) [reply frontend/public/ui/ui-actions.js]: So originally I've used the k8s-actions / k8s-reducers, but after refactoring I've put it in UI, since it's keeping data that are being rendered. 
On the other hand putting it in k8s make sense as well.

File: frontend/public/components/default-resource.jsx (hunk 3 above)
Comment (jhadvig) [reply frontend/public/components/default-resource.jsx]: Hmm.. so both options make sense to have.

>...we want a maximum number of columns to show at each screen width...

thats 12 columns for all the widths. Not sure if we want to render more then 6 columns (mainly on narrow screens) cause in case of more data the output table could get confusing.



> Alternately, we could make the table horizontally scrollable.

I like the idea, since the output would be more `oc` consistent. Could we have both ? With a `Show full table` link as a switcher?

File: frontend/public/components/default-resource.jsx (hunk 5 above)
Comment (jhadvig) [reply frontend/public/components/default-resource.jsx]: So in case of namespace we want to show the *resource-icon* infront and have the namespace as link ?

> Dates and possibly other types should be rendered specially instead of using the raw value that's returned.

So in case of **Age** we should use **Created** for header and have the output showing the same values as the ones rendered in the **Created** column (eg. Secrets table) ?
I guess other types will emerge during the work

File: frontend/public/ui/ui-actions.js (hunk 1 above)
Comment (spadgett) [reply frontend/public/ui/ui-actions.js]: I think it makes more sense under k8s. It's data coming directly from the API and that's where other watch/list data is kept.

File: frontend/public/components/default-resource.jsx (hunk 5 above)
Comment (spadgett) [reply frontend/public/components/default-resource.jsx]: The API returns a `type` property for the data. We should be looking at that to know how to format it.

**Comment by jhadvig** [comment]
@spadgett addressed comments, made some major changes and updated the PR.

After couple of rewrites I've figured out that using the table representation of the resource is not the best way how to print the resource List table in a way we do it with other resources, meaning to be sortable, there wont be conflict with columns which header has the same name as well-known headers [(name, namespace, created)](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/deploy/upstream/manifests/0.6.0/05-catalogsource.crd.yaml#L27-L30), which would make sorting implementation even more complex.

ATM the sorting takes either sortField or sortFunc. In both cases we would need to transform the flatten data array into some kind of object that the List component would be able to work with. That unfortunately would need a lot of assumptions a compromises which, after digging into the code for some time I've figured out, is just not worth, would be more complex and could cause additional problems.

I've ended up using the resource objects that the *Firehose* is returning and for knowing which columns to print I've end up using the `additionalPrinterColumns` field of the resource which CRDs have defined inside their `spec`. In case of other resources that could potentially use the default-resource list page, only *Name*, *Namespace* and *Created* columns will be printed + will be sortable.
In case of CRDs, the `additionalPrinterColumns` are containing just the additional columns, besides *Name*, so I'm adding those columns manually, if they are not added by the user in the CRD `additionalPrinterColumns` description.

One big bonus of this approach is that we have watch on the resource objects :+1: 

To sum it up the Table representation is combination of resource objects and printerColumns, but it's missing any semantics of the printed object, so we would end up just with assumptions.
I don't think the solution is perfect, since if a non-CRD resource would use the default-resource list page only *Name*, *Namespace* and *Created* columns will be printed. But from all the approaches that I've tried I definitely think it's the most efficient one.

File: frontend/public/components/chargeback.tsx [hunk 6]
```diff
@@ -36,9 +36,7 @@ import {
 } from './utils';
 
 export const ReportReference: GroupVersionKind = referenceForModel(ChargebackReportModel);
-export const ScheduledReportReference: GroupVersionKind = 'chargeback.coreos.com~v1alpha1~ScheduledReport';
```
Comment (jhadvig) [inline frontend/public/components/chargeback.tsx]: We are not using both of these anywhere.. so removing.

File: frontend/public/components/resource-pages.ts [hunk 7]
```diff
@@ -2,13 +2,13 @@
 
 import { Map as ImmutableMap } from 'immutable';
 
-import { ReportReference, ReportGenerationQueryReference } from './chargeback';
```
Comment (jhadvig) [inline frontend/public/components/resource-pages.ts]: Had to remove this one because of issue with circular deps.
Haven't come up with better solution then to remove this `import` and reinitialize `ReportReference` and `ReportGenerationQueryReference`.

File: frontend/public/module/k8s/k8s-actions.ts [hunk 8]
```diff
@@ -67,6 +72,39 @@ const actions = {
       });
   },
 
+  setPrinterColumns: (model, name) => dispatch => {
+    const kind = model.crd ? CustomResourceDefinitionModel : model;
```
[the remark this answers is not in the document]
Comment (jhadvig) [reply frontend/public/module/k8s/k8s-actions.ts]: Thought it's an API

File: frontend/public/components/default-resource.jsx [hunk 9]
```diff
@@ -16,33 +17,81 @@ import {
...
+const computeClasses = (index, columnsNumber) => {
+  const columnSize = Math.floor(12 / columnsNumber);
+  const fistColumnSize = 12 - (columnSize * (columnsNumber-1));
+  return index <= 1
+    ? `col-md-${index === 0 ? fistColumnSize : columnSize} col-sm-6 col-xs-6`
+    : `col-md-${columnSize} hidden-sm hidden-xs`;
+};
+
+const sortAndTrim = (allColumns) => {
+  const columnsSortByPriority = _.sortBy(allColumns, [function(col) {
```
[the remark this answers is not in the document]
Comment (jhadvig) [reply frontend/public/components/default-resource.jsx]: They should be

File: frontend/public/module/k8s/k8s-actions.ts [hunk 10]
```diff
@@ -67,6 +72,39 @@ const actions = {
...
+    k8sGet(kind, name).then(obj => {
```
[the remark this answers is not in the document]
Comment (jhadvig) [reply frontend/public/module/k8s/k8s-actions.ts]: > Not all users can get CRDs. I'm not sure this approach will work.

So they will be just able to list them ?

File: frontend/public/module/k8s/k8s-actions.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (spadgett) [reply frontend/public/module/k8s/k8s-actions.ts]: `model.crd` is set to true for any item that doesn't have a static model defined in the console code. But not everything it marks is really a CRD. It's going to mark many OpenShift resources as CRD for instance. (I have been meaning to rename that property.)

File: frontend/public/module/k8s/k8s-actions.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (spadgett) [reply frontend/public/module/k8s/k8s-actions.ts]: No, it looks like normal users can't list them either.

```

❯ oc --as=user@example.com get crds
No resources found.
Error from server (Forbidden): customresourcedefinitions.apiextensions.k8s.io is forbidden: User "user@example.com" cannot list customresourcedefinitions.apiextensions.k8s.io at the cluster scope: no RBAC policy matched

❯ oc adm policy who-can list crds
[...]
Groups: system:cluster-admins
        system:cluster-readers
        system:masters
```

[the remark this answers is not in the document]
Comment (jhadvig) [reply frontend/public/module/k8s/k8s-actions.ts]: so even using the Table accept header would not solve this, since its a List call and would be denied as well.
Having a link to a specific CRD in a nav section makes sense now, since the section link would be seen only to users that will be able to list that particulat CRD. right ?

[the remark this answers is not in the document]
Comment (spadgett) [reply frontend/public/module/k8s/k8s-actions.ts]: @jhadvig you can have authority to list CR instances without being able to get the CRD itself

[the remark this answers is not in the document]
Comment (jhadvig) [reply frontend/public/module/k8s/k8s-actions.ts]: @spadgett so other solution would be to try to get the CRD itself, and if that fails we could print a plain table as is with disabled sorting. if on the other hand user will be able to get the CRD we could use it to assemble the table.

Other approach that I've already tried but came to be a bit more hackish is to create our own list of resources from the table returned [table representation](https://gist.github.com/jhadvig/2d5059d9f282a1bc47fc176fc2e747ba), by converting into an array of [flattened objects](https://gist.github.com/jhadvig/5513ede2fa19fb89c49a496a079cfe9e), with which List component "could" work, although some assumptions would need to be done (eg. the [names](https://gist.github.com/jhadvig/2d5059d9f282a1bc47fc176fc2e747ba#file-table-representation-json-L10) are  not unique, ...). 
Not sure if transforming the data is the right approach but while was digging into it I found more and more problems that we would need take care of...

**Comment by spadgett** [comment]
Closing for now. We will revisit this when we have watch support for tables in k8s 1.15.
